cdap-ui/app/cdap/components/DataPrep/Directives/MapToTarget/index.js

```diff
@@ -14,36 +14,367 @@
       </div>
     );
-  }
+  };
+
+  const renderLoading = () => {
+    if (!loading) {
+      return <hr />;
+    }
+    return (
+      <div>
+        <LinearProgress />
+        <div>{T.translate(loading)}</div>
+      </div>
+    );
+  };
+
+  const renderFilter = (placeholder) => {
+    if (loading || !targetDataModel) {
+      return null;
+    }
+    return (
+      <Input
+        autoFocus={true}
+        type='text'
+        className={classes.optionSearch}
+        value={searchText}
+        placeholder={placeholder}
+        onChange={(event) => setSearchText(event.target.value)}
+      />
+    );
+  };
+
+  const renderOptions = (options, selectFn) => {
+    if (loading) {
+      return null;
+    }
+    return (
+      <List dense={true} disablePadding={true} className={classes.targetOptionList} hidden={loading}>
+        {options.map(option => (
+          <ListItem
+            button={true}
+            key={option.id}
+            id={`map-to-target-option-${option.uuid}`}
+            onClick={() => selectFn(option)}
+          >
+            <ListItemText
+              className={classes.targetOption}
+              primary={highlightText(option.name)}
+            />
+            <UncontrolledTooltip
+              target={`map-to-target-option-${option.uuid}`}
+              modifiers={{
+                preventOverflow: {
+                  boundariesElement: 'window'
+                }
+              }}
+              placement='right'
+              delay={{ show: 500, hide: 0 }}
+            >
+              {option.description || option.name}
+            </UncontrolledTooltip>
+          </ListItem>
+        ))}
+      </List>
+    );
+  };
+
+  const renderDetail = () => {
+    if (!isOpen || isDisabled) {
+      return null;
+    }
+
+    let options, selectFn;
+    let filterPlaceholder;
+    const selection = [];
+
+    if (targetDataModel) {
+      selection.push(
+        {
+          key: 'datamodel',
+          label: T.translate(`${PREFIX}.dataModelLabel`),
+          unselectFn: () => (async() => await selectTargetDataModel(null))(),
+          ...targetDataModel,
+        }
+      );
+      if (targetModel) {
+        selection.push(
+          {
+            key: 'model',
+            label: T.translate(`${PREFIX}.modelLabel`),
+            unselectFn: () => selectTargetModel(null),
+            ...targetModel,
+          }
+        );
+        filterPlaceholder = T.translate(`${PREFIX}.fieldFilterPlaceholder`);
+        options = applySearch(targetModel.fields || []);
+        selectFn = (field) => (async () => await applyDirective(field))();
+      } else {
+        filterPlaceholder = T.translate(`${PREFIX}.modelFilterPlaceholder`);
+        options = applySearch(targetDataModel.models || []);
+        selectFn = (model) => selectTargetModel(model);
+      }
+    } else {
+      options = dataModelList || [];
+      selectFn = (dataModel) => (async () => await selectTargetDataModel(dataModel))();
```

Comment:
I understand selectTargetDataModel returns a promise but do we have to await here for the promise to resolve? This is invoked from a click handler I believe. Is there a specific reason for wrapping it and making it a async function?

Reply:
Also what happens if the `setTargetDataModel` fails? Would a rejection here break the application because react unmounts the component if there is an error thrown in the render method.

Reply:
Removed `await` statements in click handlers. If `setTargetDataModel` fails then error will be displayed and component will not be unmounted. I tested it with `throw new Error()` statement.